### ISSUE TITLE
Browsers preserve fragments on redirects

### ISSUE THREAD

**Issue by domenic** [whatwg/fetch#505]
Via https://github.com/tmpvar/jsdom/pull/1762, I learned something new about browser behavior. I was wondering if this was something fetch needed to cover better? Ctrl+Fing "fragment" doesn't turn up much.

Maybe it is already handled by HTTP. But I wanted to file an issue to check.

**Comment by annevk** [comment]
They only preserve it for navigation as far as I'm aware and HTML's navigate algorithm handles it therefore.

**Comment by ylafon** [comment]
As fragments are resolved once the media type is known, HTTP nor fetch should care about preserving fragments, the logic should be left to the consumer of the media type (so defined in HTML spec for HTML).

**Comment by annevk** [comment]
@ylafon FWIW, it very much does not depend on the MIME type. HTML defines preservation for navigation to any resource, including images, plugins, media, SVG, etc.

**Comment by ylafon** [comment]
@annevk yes, I agree that there is a difference between the [theory](https://tools.ietf.org/html/rfc3986#section-3.5) and practice (HTML but also Media Fragments). Still, the fragment resolution should be defined at the consuming level (HTML), not at the protocol level.

**Comment by mnot** [comment]
Anne is correct. See http://httpwg.org/specs/rfc7231.html#header.location

**Comment by annevk** [comment]
@mnot that language suggests it would happen for subresources as well (e.g., `<img src=...#foo>` and I'm pretty sure that's not the case.

**Comment by mnot** [comment]
Possible to test it with e.g., SVG?

**Comment by domenic** [comment]
So my original concern is taken care of, feel free to close. But I'll refrain from doing so myself since it looks like maybe we've started discovering other interesting things we might want to use this to track.

**Comment by annevk** [comment]
Created a test and filed https://github.com/httpwg/http11bis/issues/37. I suppose we should update Fetch that we're actively not doing what HTTP wants here?

**Comment by annevk** [comment]
So my test confirms that browsers conform to `fetch()`, but it doesn't actually test if the fragment goes missing or if it's just not web-exposed (as Fetch requires for `fetch()` and XMLHttpRequest requires for `XMLHttpRequest`). I guess @mnot was right and we need to test it with SVG.

**Comment by annevk** [comment]
For SVG: everyone but Fx discards the fragment given in a `Location` header. Everyone respects a fragment in an "input" URL that ends up redirecting (copying it over to the final response). If both contain a fragment, only Firefox uses the one from the `Location` header, everyone else uses the fragment from the "input" URL.

So if non-Firefox browsers are willing to change, maybe we should just go with that as it matches what HTTP ended up with. We could then also change `fetch()` and `XMLHttpRequest` to start exposing the fragment (and have it exposed in service workers too).

cc @wanderview

**Comment by annevk** [comment]
@tyoshino @travisleithead @youennf any thoughts on aligning your browsers with HTTP here and propagating fragments to responses for all resources? Seems like the best way forward.

Exposing the fragments in `Response` and `XMLHttpRequest` objects is next I suppose.

I think the one more thing I can test here (if not tested already) is what `Request` reveals in service workers, to the extent that can be tested (as network redirects do not re-enter it except for navigation); and test synthetic redirect handling and fragments.

**Comment by wanderview** [comment]
> I think the one more thing I can test here (if not tested already) is what Request reveals in service workers, to the extent that can be tested (as network redirects do not re-enter it except for navigation); and test synthetic redirect handling and fragments.

I don't think that is completely true.  You can also use fetch() with "manual" mode to step-by-step follow redirects.

**Comment by travisleithead** [comment]
Hmm, I think @aliams might be able to answer better than I...

**Comment by wanderview** [comment]
Not sure if there is a separate issue for fragment in Responses, but I have a service worker related issue.

I have a bug report here:

https://bugzilla.mozilla.org/show_bug.cgi?id=1437760

Its an issue related to fragments and SVG.  The site in question just does a simple pass-through like `evt.respondWith(fetch(evt.request))`.  The request itself is for an svg resource with a fragment.  The CSS of the site looks for the fragment on the svg to determine how its rendered.

Prior to our work to propagate the synthetic Response.url back out to the intercepted network request this code would just use the initial Request.url.  We now use the final Response.url instead, though, and Response.url strips the fragment.  This breaks the CSS matching against this SVG element.

I think we need to expose the fragment on Response.url to fix this properly.  An alternative would be to try to backdoor the fragment through a private field here, but its not quiet as nice.

Does it look like we could expose the fragment on Response.url in the future?

**Comment by annevk** [comment]
Yeah, that was my plan, to propagate the fragment from the request to the response, always, and always expose it on all APIs. The main issue is getting all browsers to agree to this.

**Comment by annevk** [comment]
@tyoshino @youennf @aliams ping.

**Comment by yutakahirano** [comment]
For [fetch()](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/fetch/Response.cpp?q=Response::url&sq=package:chromium&l=343) and [XHR](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/xmlhttprequest/XMLHttpRequest.cpp?q=XMLHttpRequest::responseURL&sq=package:chromium&l=576) in Blink, fragments are just not web exposed.

I'm not familiar with the SVG behavior in Blink, @nyaxt is.

**Comment by annevk** [comment]
@yutakahirano do you think it's reasonable to start exposing them?

**Comment by yutakahirano** [comment]
What was the reason for not exposing fragments in the past?

Exposing them makes sense without considering compatibility issues, but changing the API has some risks - For example, if a hash table keyed with Response.url, may be broken with the change.

**Comment by annevk** [comment]
> What was the reason for not exposing fragments in the past?

I don't think we had a good reason and just assumed they would not have fragments because HTTP didn't really deal with them (and HTML only dealt with them for navigation). These days HTTP does deal with them and apparently some features do use them, such as SVG.

**Comment by nyaxt** [comment]
I personally think "keeping fragments" is reasonable for SVG `<use>` hrefs, but would like to hear thoughts from active SVG maintainer. @fsoder any thoughts?

**Comment by yutakahirano** [comment]
Regarding https://bugzilla.mozilla.org/show_bug.cgi?id=1437760, can you tell me where [#dom-response-url](https://fetch.spec.whatwg.org/#dom-response-url) is used?  I'm under impression that [responseWith](https://w3c.github.io/ServiceWorker/#dom-fetchevent-respondwith) copies the concept response and hence doesn't care about [#dom-response-url](https://fetch.spec.whatwg.org/#dom-response-url).

**Comment by wanderview** [comment]
@yutakahirano Well, `respondWith` simply doesn't overwrite the `response.url` with `request.url` any more.  I couldn't figure out how the SVG code actually accesses the url from the response.  Arguably it could use the internal url representation, but its unclear to me.

Currently in gecko, though, we don't preserve the fragment in the Response internally because it would require extra work to properly store it in Cache API, etc.  It sounds like we will need to add that, but currently we don't.

Besides that specific bug, though, it does seem reasonable to at least apply the request URL fragment to the response URL since its what browsers do for redirects.

Note, I believe the WPT reftest I added in https://bugzilla.mozilla.org/show_bug.cgi?id=1437760 would be satisfied by any of these implementations:

* legacy behavior of overwriting the response.url with the request.url
* propagating the internal response url with a fragment
* applying the request.url fragment to the response.url

So I hope that test is not controversial.

**Comment by fsoder** [comment]
@nyaxt not sure what "keeping fragments" refer exactly, but fragments carry relevance in many contexts for SVG (resource references - including `<use>` - as well as image contexts [`<img>` and CSS `<image>` values et.c].) In Blink we're not really setup to use a fragment (if any) from the response ATM.

**Comment by hober** [comment]
cc @achristensen07

**Comment by annevk** [comment]
This also affects `XMLHttpRequest`'s `responseURL` which we should probably fix as well for consistency.

**Comment by yutakahirano** [comment]
I still don't understand why fixing the url getter of Response class is needed.

**Comment by annevk** [comment]
@yutakahirano how else would a web developer implement SVG-like behavior themselves?

**Comment by yutakahirano** [comment]
@annevk does that mean the only argument is API completeness and the change is not related to existing problems (such as navigation related one mentioned in the thread), right? I'm not opposing but want to understand the motivation.

**Comment by annevk** [comment]
@yutakahirano yeah, exposing it in the API is more about completeness and exposing the underlying primitive upon which the rest of the things build. (It does seem like some of this will require implementation changes as not everyone forwards fragments today per my earlier testing which I'll turn into web-platform-tests.)

**Comment by yutakahirano** [comment]
I see. I can roughly measure how many pages are affected. It will take time (2-3 months) though.

I guess that not many pages will be broken. Assuming that, if other browser vendors are fine with the idea I can change the behavior.

**Comment by annevk** [comment]
I wrote tests in https://github.com/w3c/web-platform-tests/pull/7261 and a PR for Fetch and XMLHttpRequest. I couldn't really find existing tests for this.

**Comment by hiroshige-g** [comment]
> propagate the fragment from the request to the response, always

How this works when two requests with different fragments are served by the same cache entry?
For example,

```
<img src="/images/green.svg#svgView(viewBox(0,0,50,50))">
<img src="/images/green.svg#svgView(viewBox(25,25,50,50))">
```

Are two response objects with URLs with different fragments are created in Step 20.1 of https://fetch.spec.whatwg.org/#http-network-or-cache-fetch?

**Comment by annevk** [comment]
@hiroshige-g no, the response URL is set at a later point (step 9 of main fetch) and shouldn't affect the HTTP cache.

**Comment by dcleao** [comment]
@annevk could you please clarify your above comment about fragments (https://github.com/whatwg/fetch/issues/505#issuecomment-366671504), where you say:
> These days HTTP does deal with them

I couldn't find a spec stating that fragments are used for HTTP caching. Or, are you referring to some other HTTP concern?

Thanks in advance!

**Comment by annevk** [comment]
This issue is about whether to include them in redirects.

**Comment by annevk** [comment]
@davidben I'm wondering if you have any thoughts how to best account for fragments in redirects. It seems that when computing the location URL we'd really want the request's current URL's fragment as an input to that algorithm. An alternative is that we handle setting fragment separately, as this PR does now, but that would require handling it separately in HTML's navigate algorithm too.

**Comment by annevk** [comment]
@wanderview @jakearchibald @asuth @youennf I'd like to finally solve fragments in URLs and service workers. As far as I can tell we have these scenarios (not sure if synthetic really matters):

```
Document	Service worker responds with	Outcome
Fetch x		Synthetic					x
Fetch x		Fetch y						y
Fetch x		Fetch y#y					y???
Fetch x#x	Synthetic					x#x
Fetch x#x	Fetch y						y#x
Fetch x#x	Fetch y#y					y???
```
Are there scenarios I missed here?

What I'm observing in Firefox (I find it hard to test in Chrome and Safari due to certificate overrides being a bit of a pain) is that request's fragment is everything. So the first `y???` would be `x` and the second would be `y#x`. This is different from what #696 has (which is the PR from 2018 (that I think got abandoned due to me forgetting about it after vacation) rebased) in that the response is not authoritative when it comes to the fragment. Which I think is a bit weird with how we generally treat service worker responses and the fact that with redirects the location URL fragment is authoritative if present. I no longer care strongly though so I'll defer to you on what to specify here.

**Comment by wanderview** [comment]
FWIW, this is where firefox does the fragment fixup:

https://searchfox.org/mozilla-central/rev/3ff133d19f87da2ba01ade55d6e7fdf0fc28b08c/dom/serviceworkers/ServiceWorkerEvents.cpp#704-715

The rationale for this is discussed previously in this thread here:

https://github.com/whatwg/fetch/issues/505#issuecomment-365373835

So I believe your question mark outcomes would be `y` and `y#x`.  (See the assert that the response URL never has a fragment exposed.)

**Comment by wanderview** [comment]
Also note the comment that if response.url is changed to include the fragment then we would avoid overriding in that case.  So the final example you have would become `y#y` instead.

**Comment by annevk** [comment]
Thanks, that was a typo. Any thoughts on what we should be doing here?

**Comment by davidben** [comment]
Oh fun. I hadn't thought of that. I haven't had a chance to actually try it, but looking at the code, our current implementation is in this `ComputeRedirectInfo` function:
https://source.chromium.org/chromium/chromium/src/+/master:net/url_request/redirect_info.cc;l=137-149;drc=2ac64302ae161cd6b5e4b1254497bdf5fd6d3415;bpv=1;bpt=0

Which ends up copying the fragment from the request's URL *after* we've already resolved the location URL to `new_location`. That's called both in the net stack (where this isn't so messy) and when we follow a Service-Worker-provided fragment. So I think that means our current behavior is:

* Response URL's fragment is irrelevant. It is the base URL when resolving the location URL, but fragments of base URLs don't matter.
* If the result of `response_url.Resolve(location_header)` has no fragment and the request URL has a fragment, use the request URL's fragment.

If I understand you correctly, it sounds like Chrome and Firefox have the same behavior right now.

I suspect this behavior is the right one: consider any alternative where we used the response.url fragment (i.e. give back y#y). That would require the response URL preserve the fragment which, in turn, means `fetch("foo")` and `fetch("foo#bar")` give different responses. That would then mess up caching and probably is not what we want.

Though this does probably mean having to duplicate the fragment copying in HTTP-redirect fetch and navigation. (Or does it? Do fragments matter outside navigation?)

> Also note the comment that if response.url is changed to include the fragment then we would avoid overriding in that case. So the final example you have would become y#y instead.

I guess this is a question of whether you believe resolving the Location header against response.url runs the "parse URL" algorithm or "parse URL" followed by "copy over fragment for redirects". Relative URL resolution normally ignores the base URL's fragment.

**Comment by annevk** [comment]
Your comment made me remember that response's URL (list) is just a copy of request's URL (list), and when it's not, it's null. So I think that means that in this case we can actually use the fragment from the base URL (the response's URL). Which means we don't have to pass an additional parameter to the location URL algorithm and that also means HTML does not need changes. And yes, it's ignored during parsing `Location`'s value, but [HTTP requires](https://httpwg.org/specs/rfc7231.html#header.location) that it's still used if the resulting location URL does not have a fragment. I think that settles it for redirects.

Now, Chrome/Safari do not have the same behavior as Firefox, see https://github.com/web-platform-tests/wpt/pull/27575. Firefox has the behavior I described for all redirects, but Chrome/Safari only do it for navigation. Per HTTP that is a bug and back in 2018 we reached agreement on aligning with HTTP.

I think for redirects that's the right approach and I will update the PR accordingly.

Now for service workers it's less clear. I agree that any caches should not store fragments and ignore them when they are used in a key. So if the service worker returns something from the Cache API it would always be equivalent to the responding with Fetch `y` scenario and never `y#y`. But when the service worker fetches itself and responds with that it's less clear to me. In particular because `y` wins over `x`. Although we do know that `#x` needs to win over nothing. So maybe we should treat the service worker as being able to supply us with responses and those responses keep their own URL identity, but only up to the fragment, which is always set by the request, be it to null or a string. (Which coincidentally would align with implementations today.)

**Comment by wanderview** [comment]
>  I agree that any caches should not store fragments and ignore them when they are used in a key.

I understand ignoring them in the request url key, but why would we not preserve them on response.url when the response is stored in cache_storage?  I didn't follow that part.

> So maybe we should treat the service worker as being able to supply us with responses and those responses keep their own URL identity, but only up to the fragment, which is always set by the request, be it to null or a string.

I think this would miss some use cases.

Consider the use case from https://github.com/whatwg/fetch/issues/505#issuecomment-365373835.  We have a request to an svg with multiple sprites in it and the fragment identifies which one to display in an `<img>`.  Consider that the service worker wants to return a different svg with a different set of sprites.  The anchors in this new svg may not map to the anchors in the original svg.  The service worker should be able to identify a different sprite within its chosen svg with a new/different fragment.

**Comment by annevk** [comment]
It seems to me that if you store them in the cache and the service worker is authoritative you might end up with accidental fragments there that will then forever be returned?

I'm okay with making the fragment authoritative if it comes from the service worker though. And for the cache case we could let the cache.match request URL decide?

**Comment by wanderview** [comment]
I don't think the cache.match request URL should be involved at all.  The state of the response is orthogonal to the key its stored under.

The principle I'm trying to preserve is that cache_storage does not mutate responses that flow through it.  So that this:

```javascript
evt.respondWith(fetch(request));
```

Should be equivalent to this:

```javascript
evt.respondWith(async function() {
  const r = await fetch(request);
  const c = await caches.open(name);
  c.put(request, r);
  return c.match(request);
}());
```

If we make cache_storage behave differently than a response returned from fetch() it creates a barrier to sites trying to use cache_storage for offline, reliability, etc.  Therefore I feel cache_storage should preserve the entire response exactly and reproduce it exactly (as possible).

To reiterate my overall desired outcome:

* If response.url exists and has a fragment, then respondWith() propagates that full url to the outer response.
* If response.url exists, but does not have a fragment, the respondWith() appends any FetchEvent.request.url fragment to the response.url.
* If response.url does not exist, then FetchEvent.request.url is uses.
* Response.url is preserved in cache_storage with fragments.

**Comment by annevk** [comment]
Again though, say you navigate to `x#x`, the service worker caches it and returns it. Then you navigate to `x` in a new tab. It'll get `x#x` back. Is that really what developers would expect?

**Comment by wanderview** [comment]
Navigations are different.  They ignore the response.url returned from service worker.  As I recall this was an early design goal so a service worker could return an "offline page" without changing the URL bar.  I don't know where the spec does that in navigation-land, though.

**Comment by annevk** [comment]
Okay, but doesn't the same concern apply to subresources? Document A has an SVG with `#viewBox(blah)`, document B has one with `#viewBox(blahblah)`? It seems that A would poison the cache quite easily.

**Comment by wanderview** [comment]
Hmm, I see your point.  Thanks for helping me understand.  It seems this could even happen within a single document if they want to pull different sprites out of the same svg (using my example from above).

I guess sw's that care about this could always strip the fragment themselves, but that seems non-ergonomic.  We could provide an option to auto-strip at put() or match() time.  Or an option to opt-out of auto-strip at put() or match() time.  Or, as you suggest, provide a way to specify the fragment at match() time either through request url or an option.

Or go with your recommendation to always use the FetchEvent.request.url fragment if present.

Keeping the status quo of the FetchEvent.request.url fragment seems maybe best for now.  If we do that, though, the fragment would be preserved on cache_storage response.url?

**Comment by annevk** [comment]
Assuming that you mean that request's URL's fragment would always win, even when null, I think we should decide that final question as part of #1167. It seems a little weird to me to store it, but I could be convinced I suppose.

**Comment by wanderview** [comment]
Ok, I see what you mean about using FetchEvent.request.url fragment whether its set or null.  That would be the safest change I guess.

But I still feel strongly that cache_storage should not mutate values stored in it.

**Comment by asutherland** [comment]
Is there an existing rationale on making it so that the Response constructor doesn't allow content to specify/override a URL and thereby add/drop the fragment?  Right now the operations available to content script are:
- Forge a URL on a Response in a SW client (but not in a SW) by asking for whatever it wants via fetch() and having the SW respondWith a synthesized Response via Response constructor which will not have a URL.  This is part of the principle that an origin can lie to itself (but currently in limited ways).
  - Because Responses are not Serializable, this can't be propagated via postMessage.
- Drop the URL off a (non-opaque) Response by re-synthesizing it via `new Response(oldResponse.body, oldResponse)`.

If we're planning on making Responses Serializable in the future, it might make sense to make it more possible to control the URL of Responses more directly?

**Comment by wanderview** [comment]
> Because Responses are not Serializable, this can't be propagated via postMessage.

It can be propagated via cache_storage today.

**Comment by asutherland** [comment]
Oh right, good points, thanks!  I've added a note to my point inline to avoid propagating confusion about that.

**Comment by annevk** [comment]
I could maybe see allowing developers to set same-origin URLs. I'm not sure I see that changing our options with regards to fragment handling though. (Have there been requests for that? Should probably be a new issue.)

**Comment by wanderview** [comment]
After reviewing the WPT tests I kind of feel like we should match non-sw redirect behavior here.  So a fragment in the SW-provided response.url would override the request.url.  Cache pollution would be a potential problem, but sites can avoid that by storing responses without the fragment if they need to.  (They can purposely strip the request fragment when fetching the response to store in cache_storage, etc.) 

Not sure this is a hill I'm willing to die on, but I'm leaning more towards that solution than having redirects and respondWith() diverge in behavior.

**Comment by annevk** [comment]
@wanderview it seems you didn't engage in @davidben's feedback on the PR, who suggested a different approach that's not vulnerable to cache pollution. Could you please review that? Meanwhile, I think I'm going to drop the service worker changes and move those to a new issue as it seems those will take longer to reach agreement on and get the correct tests for.

**Comment by wanderview** [comment]
Can you link to David's comment?  There have been a number of PRs and issues flying around and I can't find it.  Sorry.

**Comment by wanderview** [comment]
Nevermind I found it.

**Comment by annevk** [comment]
I linked it all in #1175 which will be the service worker and fragments issue going forward. This issue is now limited to redirects and hopefully addressed by #696.
